docs: add new benchmark page with SIFT1M dataset (#691)
docarray/docarray: docs/advanced/document-store/benchmark.md

:scale: 0 %
```

> We create a DocumentArray with one million Documents based on [SIFT1M](https://www.tensorflow.org/datasets/catalog/sift1m), a dataset containing 1 million objects of 128d and using l2 distance metrics, and benchmark the following document stores.

Reviewer: I'm not sure whether the link from tensorflow is the official sift1m page I think the home page is http://corpus-texmex.irisa.fr/

| Name                                | Usage                                    | Client Version    | Database Version |
|-------------------------------------|------------------------------------------|-------------------|------------------|
| In-memory DocumentArray             | `DocumentArray()`                        | DocArray `0.18.2` | N/A              |
| [`Sqlite`](sqlite.md)               | `DocumentArray(storage='sqlite')`        | `2.6.0`           | N/A              |
| [`Weaviate`](weaviate.md)           | `DocumentArray(storage='weaviate')`      | `3.3.3`           | `1.13.2`         |
| [`Qdrant`](qdrant.md)               | `DocumentArray(storage='qdrant')`        | `0.8.0`           | `0.8.0`          |
| [`Annlite`](annlite.md)             | `DocumentArray(storage='anlite')`        | `0.3.13`          | N/A              |
| [`ElasticSearch`](elasticsearch.md) | `DocumentArray(storage='elasticsearch')` | `8.4.3`           | `8.2.0`          |
| [`Redis`](redis.md)                 | `DocumentArray(storage='redis')`         | `4.3.4`           | `2.6.0`          |

We focus on the following tasks:

> 

Reviewer: and for the filter operation, let's say it uses backend side filtering when possible

> 

Reviewer: in the text below, instead of saying:

> We use similar underlying search algorithms but different implementations**: In this benchmark we focus on setting only parameters `ef`, `ef_construct` and `max_connections` from HNSW

let's say

> We use similar underlying search algorithms but different implementations**: In this benchmark we try a set of parameters `ef`, `ef_construct` and `max_connections` from HNSW applied equally on all backends.

:scale: 0 %
```

We create a DocumentArray with one million Documents based on [sift1m](http://corpus-texmex.irisa.fr/), a dataset containing 1 million objects of 128d and using l2 distance metrics, and benchmark the document stores summarized below.

This includes classic database and vector database, all under the same DocumentArray API:

| Name                                | Usage                                    | Client Version    | Database Version |
|-------------------------------------|------------------------------------------|-------------------|------------------|
| In-memory DocumentArray             | `DocumentArray()`                        | DocArray `0.18.2` | N/A              |
| [`Sqlite`](sqlite.md)               | `DocumentArray(storage='sqlite')`        | `2.6.0`           | N/A              |

> | [`Weaviate`](weaviate.md)           | `DocumentArray(storage='weaviate')`      | `3.3.3`           | `1.13.2`         |

Reviewer: update weaviate version

> | [`Qdrant`](qdrant.md)               | `DocumentArray(storage='qdrant')`        | `0.8.0`           | `0.8.0`          |

Reviewer: update qdrant version

* **Benchmark is for DocArray users, not for research**: This benchmark showcases what a user can expect to get from DocArray without tuning hyper-parameters of a vector database. In practice, we strongly recommend tuning them to achieve high quality results.
```

## Benchmark result

The following chart and table summarize the result. The chart depicts Recall@10 (the fraction of true nearest neighbors found, on average over all queries) against QPS (find by vecotr queries per second). The smaller the time values and the more upper-right in the chart, the better.

## Charts

```{include} benchmark.html
```

> ````{tab} None

Reviewer: let's change None

|              32 |          256 |   64 |     0.984 |                1.62 |                   0.37 |        211.81 |      0.35 |       32.65 |       35.15 |
|              32 |          256 |  128 |     0.996 |                1.84 |                   0.37 |        210.85 |      0.35 |      141.31 |      175.80 |
|              32 |          256 |  256 |     0.999 |                2.25 |                   0.37 |        204.65 |      0.35 |       22.13 |       31.54 |

````

````{tab} Qdrant

| max connections | ef construct |  ef  | Recall@10 | Find by vector (ms) | Find by condition (ms) | Create 1M (s) | Read (ms) | Update (ms) | Delete (ms) |
|-----------------|-------------:|-----:|----------:|--------------------:|-----------------------:|--------------:|----------:|------------:|------------:|
|              16 |           64 |   32 |     0.968 |               47.80 |                  44.96 |       5889.66 |      5.70 |       46.11 |       47.24 |
|              16 |           64 |   64 |     0.990 |               47.88 |                  44.72 |       5892.35 |      2.72 |       46.61 |       47.30 |

> |              16 |           64 |  128 |     0.997 |               47.80 |                  44.95 |       5881.15 |      2.84 |       45.34 |       47.36 |

Reviewer: higher recall as the ef value increase implies that the configuration is correctly applied, although find by vector stays constant.
I doubt big network overhead here, I'm trying to confirm

|              32 |          128 |   64 |     0.975 |                2.10 |                   0.67 |        953.10 |      1.06 |        2.35 |       27.11 |
|              32 |          128 |  128 |     0.993 |                2.49 |                   0.69 |        921.87 |      1.03 |        3.06 |       27.58 |
|              32 |          128 |  256 |     0.998 |                3.06 |                   0.64 |        926.96 |      1.06 |        2.45 |       27.27 |
|              32 |          256 |   64 |     0.984 |                2.28 |                   0.79 |       1489.83 |      1.05 |        4.92 |       29.27 |
|              32 |          256 |  128 |     0.996 |                2.75 |                   0.79 |       1511.17 |      1.05 |        4.03 |       28.48 |
|              32 |          256 |  256 |     0.999 |                3.15 |                   0.63 |       1534.68 |      1.03 |        3.26 |       28.19 |

````

When we consider each query as a Document, we can convert the above metrics into query/document per second, i.e. QPS/DPS. Values are higher the better (except for `Recall@10`). 

> ````{tab} None in QPS

Reviewer: change None

|  32 |          128 |  128 |     0.993 |                2.49 |                   0.69 |        921.87 |      1.03 |        3.06 |       27.58 |
|  32 |          128 |  256 |     0.998 |                3.06 |                   0.64 |        926.96 |      1.06 |        2.45 |       27.27 |
|  32 |          256 |   64 |     0.984 |                2.28 |                   0.79 |       1489.83 |      1.05 |        4.92 |       29.27 |
|  32 |          256 |  128 |     0.996 |                2.75 |                   0.79 |       1511.17 |      1.05 |        4.03 |       28.48 |
|  32 |          256 |  256 |     0.999 |                3.15 |                   0.63 |       1534.68 |      1.03 |        3.26 |       28.19 |

````

````{tab} Sqlite

|  m  | ef_construct |  ef  | Recall@10 | Find by vector (s) | Find by condition (s) | Create 1M (s) | Read (ms) | Update (ms) | Delete (s) |
|-----|-------------:|-----:|----------:|-------------------:|----------------------:|--------------:|----------:|------------:|-----------:|
| N/A |          N/A |  N/A |     1.000 |              54.32 |                 78.63 |     16,421.51 |      1.09 |       0.40  |      28.87 |

````

> When we consider each query as a Document, we can convert the above metrics into query/document per second, i.e. QPS/DPS. Higher values are better (except for `Recall@10`). 

Reviewer: > except for `Recall@10`

did you mean something else ?
because also high recall is good

> * **Not just speed, but also quality**: Since some backends offer better performance and some offer better quality, we make sure to report the quality measurement for the approximate nearest neighbor search. This will allow users to 

Reviewer: I think we should say here that we show the trade-off between quality and speed as you tune your parameters in each document store

choose the backend that best suits their cases.
* **Same experiment, same API**: DocArray offers the same API across all backends and therefore we built on top of it the 
same benchmarking experiment. Furthermore, we made sure to run the experiment with a series of HNSW parameters for backends that support 
approximate nearest neighbor search. All backends are run on official Docker containers, local to the DocArray client 
which allows having similar network overhead. We also allocate the same resources for those Docker containers and all 

> servers are run in a single node setup.

Reviewer: I'm referring to  the text below:

> [raise a PR or issue](https://github.com/jina-ai/docarray/issues/new/choose).
let's change this link to the new docarray repository

If your dataset does not fit in memory, and you **do not** care much about the speed of nearest neighbor search, you can use `sqlite` as storage.

```{tip}
SQLite store is omitted because SQLite is too slow to fit into the figure chart.
```

> AnnLite is a good choice when indexing/appending/inserting speed matters more than the speed of finding. Moreover, AnnLite is a local monolithic package that does not follow a client-server design, so it avoids all network overhead.

Reviewer: ```suggestion
AnnLite offers good speed in CRUD and vector search operations, but keep in mind that AnnLite is a library and does not follow a client-server design. This means that AnnLite operations do not include network overhead unlike other backends.

Moreover, AnnLite is a local monolithic package and does not offer scaling. However, you can still scale using [Jina's scaling features](https://docs.jina.ai/how-to/scale-out/).

```

> If your dataset does not fit in memory, and you **do not** care much about the speed of nearest neighbor search, you can use `sqlite` as storage.

Reviewer: we still don't talk about other backends. I think we can share the following insight:
* Redis offers great speed if your data fits in memory and you need an isolated server to store your data

```{tip}

> SQLite store is omitted because SQLite is too slow to fit into the figure chart.

Reviewer: this belongs to the removed chart, we should also remove this

```

AnnLite offers good speed in CRUD and vector search operations, but keep in mind that AnnLite is a library and does not follow a client-server design. This means that AnnLite operations do not include network overhead unlike other backends.

> Moreover, AnnLite is a local monolithic package and does not offer scaling. However, you can still scale using [Jina's scaling features](https://docs.jina.ai/how-to/scale-out/).

Reviewer: ```suggestion

Moreover, AnnLite is a local monolithic package and does not offer scaling. However, you can still scale using [Jina's scaling features](https://docs.jina.ai/how-to/scale-out/).
```

GRPC instead of HTTP, or use batch operations. We're okay with that, as long as it benefits the DocArray and Jina 
user.
* **Open to improvements**: We are constantly improving the performance of storage backends from the DocArray side and 
updating the benchmarks accordingly. If you believe we missed an optimization (e.g. perform an operation in batches, benefit from a recent feature in upstream, avoid unnecessary steps), feel free to [raise a PR or issue](https://github.com/docarray/docarray). We're open to your contributions!

### Incompleteness on the stores

We do not benchmark algorithms or ANN libraries like **Faiss, Annoy, Scann**. We only benchmark backends that can be used as 
Document stores. Actually we do not benchmark HNSW itself, but it is used by some backends internally.

Other storage backends that support vector search are not integrated with DocArray yet. We're open to contributions to DocArray's repository to support them.

> 

Reviewer: ```suggestion
#### Client/Server setup
```
Let's say here that in our benchmarks the client and server live on the same machine to minimize network overhead.
We need to say that although a real life scenario would be clients and server living  on different machines with potentially multiple clients in parallel, we choose to keep both in the same machine and have only 1 client process
